### PR TITLE
Use DOM manipulation to create HTML

### LINKED ISSUE #122: JavaScript: HTML not built with DOM manipulation
WordPress VIP support will not allow this plugin to be used until this issue is addressed:

## js/rpbchess-ui-chessboard.js
* Line 298, HTML should be built using DOM manipulation when using arbitrary variables. See https://vip.wordpress.com/2015/03/25/preventing-xss-in-javascript/
* Line 313 (`buildContent` function), HTML should be built using DOM manipulation when using arbitrary variables. See https://vip.wordpress.com/2015/03/25/preventing-xss-in-javascript/
* Line 975, HTML should be built using DOM manipulation when using arbitrary variables. See https://vip.wordpress.com/2015/03/25/preventing-xss-in-javascript/

## js/rpbchess-ui-chessgame.js
* Line 425, HTML should be built using DOM manipulation when using arbitrary variables. See https://vip.wordpress.com/2015/03/25/preventing-xss-in-javascript/
* Line 605 (`refresh` function), HTML should be built using DOM manipulation when using arbitrary variables. See https://vip.wordpress.com/2015/03/25/preventing-xss-in-javascript/
* Line 799 (`buildErrorMessage` function), HTML should be built using DOM manipulation when using arbitrary variables. See https://vip.wordpress.com/2015/03/25/preventing-xss-in-javascript/
* Line 873 (`playerNameHeader` function), HTML should be built using DOM manipulation when using arbitrary variables. See https://vip.wordpress.com/2015/03/25/preventing-xss-in-javascript/
* Line 899 (`datePlaceHeader` function), HTML should be built using DOM manipulation when using arbitrary variables. See https://vip.wordpress.com/2015/03/25/preventing-xss-in-javascript/
* Line 923 (`annotatorHeader` function), HTML should be built using DOM manipulation when using arbitrary variables. See https://vip.wordpress.com/2015/03/25/preventing-xss-in-javascript/

### PR DESCRIPTION
**Pull request by paulschreiber**
Fixes a security hole when HTML is created by string concatenation.

Fix #122. 

(Do not close yet — so far, 3/9 functions have been fixed.)

### CODE REVIEW

File: js/rpbchess-ui-chessgame.js [hunk 1]
```diff
@@ -1102,10 +1242,20 @@
...
+		var result = document.createElement( tag );
+		result.setAttribute( 'class', 'rpbui-chessgame-comment' );
+
+		var positionInformation = buildPositionInformation( node, false );
+		for ( var key in positionInformation ) {
+			if ( positionInformation.hasOwnProperty( key )) {
+				result.setAttribute( key, value );
+			}
+		}
+		result.textContent = node.comment();
```
Comment (yo35) [inline js/rpbchess-ui-chessgame.js]: `node.comment()` may contain some HTML tags, to enhance comment rendering. With your modification, this feature will be lost.

File: js/rpbchess-ui-chessgame.js [hunk 2]
```diff
@@ -869,23 +972,30 @@
...
+		var round = formatDefault( widget._game.header( 'Round' ) );
 
-		// Build and return the header.
-		var header = '<div class="rpbui-chessgame-event">' + event;
-		if(round !== null) {
-			header += '<span class="rpbui-chessgame-round">' + round + '</span>';
+		// Build and return the header
+		var header = document.createElement( 'div' );
+		header.setAttribute( 'class', 'rpbui-chessgame-event' );
+		header.textContent = event;
```
Comment (yo35) [inline js/rpbchess-ui-chessgame.js]: The headers may contain some HTML tag, to enhance rendering. With your modification, this feature will be lost.

There is an example of this feature in the admin > 'Chess' > 'Memo':
code

**Comment by yo35** [comment]
PS: besides this two problems, I don't really see the point in using direct DOM manipulation instead of the jQuery API. This leads to a code that is much longer (+492 −223 lines of code in your PR).

**Comment by paulschreiber** [comment]
If you want to do this with the jQuery API, that's fine, too. The problem is using string concatenation, where HTML fragments (class names, tags, contents) come from variables. The could means to XSS vulnerabilities.

**Comment by paulschreiber** [comment]
I've fixed the two issues noted above.

**Comment by yo35** [comment]
You've fixed the two issues I mentioned by replacing `.textContent = ...;` with `.innerHTML = ...;`. But as far as I understand, this change re-opens the security hole, by allowing code injection.

From the article you mentioned:

> To avoid XSS, we want to avoid inserting HTML directly into the document and instead, programmatically create DOM nodes and append them to the DOM. This means avoiding .html(), .innerHTML, ...

**Comment by paulschreiber** [comment]
Good point regarding `.innerHTML`. I've reverted that comment. The best way to include HTML would be to use `.appendChild()`.

However, I'm not sure where the content for `comment()` and `event` come from. It is from user input?

**Comment by yo35** [comment]
The content for `comment()` and `event` may come from user input, more precisely it comes from post/page text located between `[pgn]...[/pgn]` shortcodes. There is an example of how this feature can be used in the admin > 'Chess' > 'Help' > 'PGN game syntax' > 'Comments'

```
[pgn]
1. d4 { I'm a <strong style="color:red;">short</strong> comment. } 1... Nf6 2. c4 e6
...
[/pgn]
```
The text between braces in the comment.

**Comment by paulschreiber** [comment]
I'm not sure how you'd sanitize the user input to make sure it's safe.

Code-wise, you could do this:
```javascript
var eventNode = $('<span>' + event + '</span>')[0];
header.textContent.appendChild( eventNode );
```

Which gets rid of the `.innerHTML` call. You're still vulnerable to XSS if `event` contains dangerous code.

**Comment by paulschreiber** [comment]
Do you have any suggestions on how to fix this? We'd like to include the Chess plugin. This is the last blocker.

**Comment by yo35** [comment]
This is also the hardest blocker! ;-) The problem is that we DO want to allow HTML tags in PGN comments -- or at least some HTML tags, typically those aiming at customizing font (`<strong>`, `<em>`, `<span>`...).
Typically, I'll have to implement some kind HTML validation and filtering, possibly using https://www.npmjs.com/package/striptags. Would it be OK on your side?

**Comment by yo35** [comment]
As explained in #122, I take the opportunity of this fix to achieve a more significant refactoring. However, thanks for your inspiring work.

**Comment by paulschreiber** [comment]
I'd recommend that you sanitize on the server side as well. Using `wp_kses()` or `wp_kses_post()` will ensure anything written to the database is sane.

**Pull request closed without merging**